example_allocators/compartment_alloc/compartment_alloc.c

```diff
@@ -0,0 +1,91 @@
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include "compartment_alloc.h"
+#include "utils.h"
+
+heap_compartment compartments[maxCompartments];
+int numCompartments = 0;
+
+/*
+ * Initialise a compartment with `size_in_bytes` space, where all future
+ * allocations will be constrained by `dc` as if it was the DDC when allocation
+ * occurred.
+ */
+void
+*__capability init_compartment(int size_in_bytes, void *__capability dc)
+{
+    // We impose a limit on our # of compartments
+    if (numCompartments + 1 == maxCompartments) {
+        return NULL;
+    }
+
+    // Allocate memory for this compartment
+    void *buf = mmap(NULL, size_in_bytes, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+    if (buf == MAP_FAILED)
+    {
+        perror("error in initial mem allocation");
+        exit(-1);
+    }
+
+    // Create the new compartment
+    compartments[numCompartments] = (heap_compartment){
+        buf,                     // Space on the heap
+        0,                       // bytes allocated (currently none!)
+        size_in_bytes,           // Our maximum size, to check when allocating
+        dc,                      // The data capability to allocate against (like a DDC)
+        sealed_reference(buf)    // An identifier which allows a holder to allocate in this component
+    };
+
+    // Return the sealed capability which acts as a token of authentication for a component.
+    return compartments[numCompartments++].identifier;
+}
+
+/*
+ * Allocate `len` bytes in the compartment identified by `compartment_id`.
+ * if `compartment_id` doesn't exist, returns NULL
+ */
+void *__capability malloc_compartment(int len, void *__capability compartment_id)
+{
+    size_t rounded_length;
+    size_t new_allocated;
+    heap_compartment compartment = compartments[0];
+    void *__capability allocated;
+    void *addr;
+    int i;
+
+    // Search for a compartment with the given identifier
+    for (i = 0; i < maxCompartments && compartments[i].identifier != compartment_id; i++){}
+    if (i == maxCompartments)
+    {
+        perror("Given an ID for a non-existent compartment");
+        return NULL;
+    }
+
+
+    compartment = compartments[i];
+
+    // Try to "bump-allocate" some space in the compartment's buffer
+    addr = compartment.buffer + compartment.bytes_allocated;
+
+    addr = __builtin_align_up(addr, ~cheri_representable_alignment_mask(len) + 1);
+    rounded_length = cheri_representable_length(len);
+
+    new_allocated = (addr + rounded_length) - compartment.buffer;
+    if (new_allocated > compartment.max_allocated) {
+        return 0;
+    }
+
+    compartment.bytes_allocated = new_allocated;
+
+    // We allocated some space!
+    // Create a capability pointing to it and return it.
+    // The capability is bounded by the data capability for this compartment.
+    allocated = cheri_cap_build((void *__capability)addr, (__uintcap_t)compartment.datacap);
```

Comment:
I must admit that I don't think I've seen `cheri_cap_build` before -- I think it's new? Though Yury has used it in his blog post, I see!

Reply:
Is there some documentation somewhere for what this function is meant to do? It might be best to just cast to a capability to avoid any potential side-effects if we're not sure what this is meant to accomplish.

Reply:
It's intended to create a capability with the metadata of `compartment.datacap` (bounds, permissions etc) and the address of `addr`. For some reason it requires the address as a capability and doesn't seem to set it properly in this example…I've got a snippet where it works as expected and never solved the mystery here. It needs tidying up. I didn't realise it was uncommon, I'll change this to be more idiomatic!

The result I wanted to achieve with these three lines was to have a capability with the address `addr`, bounds set to exactly the bounds of the allocated # bytes, and all other data derived from `compartment.datacap` (I set r/w permissions in `main.c`, for example). Is there an idiomatic way to construct that capability without using `cheri_cap_build()`?

Reply:
Can you use something like `cheri_address_set(compartment.datacap, addr)`? Warning: I might be wrong.

Reply:
I would just cast to `__capability` (and then you'll need `__from_cap` I think - you should get a warning) instead of `cheri_cap_build`, which I guess we don't know where it comes from?

Reply:
Moved to `cheri_address_set` in [3c39dc7](https://github.com/capablevms/cheri-examples/pull/88/commits/3c39dc77ae584f060b5cc633c019f3979695adf4)

Reply:
We did some investigation around `cheri_cap_build` and, in the initial usage, it was a no-op. We will likely have an example to submit with `cheri_cap_build` usage, which @probablytom will take care of, just to show how it can be used. We also found a comment from @jacobbramley that explains how this function should be used (as documentation on it seems severely lacking).

[1] https://github.com/capablevms/cheri-examples/pull/39#discussion_r714883024